From be1bac0463b4c752760f235b6f4e3b6bd005c4d6 Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Fri, 16 Jan 2015 18:42:08 +0800
Subject: [PATCH 0403/5242] MLK-10103-1 ARM: imx: add VPU 352M for i.mx6q

commit  c6f2770765bc5a45f449bf9517f39d3a19a8c79d from
https://source.codeaurora.org/external/imx/linux-imx.git

When VPU freq is set to 352MHz, it needs to source clk
from PLL2_PFD2_396M, and PLL2_PFD2_396M need to change
freq to 352M, cpufreq's 396M setpoint will be removed.

Busfreq will be disabled as it needs PLL2_PFD2 to be
as 396MHz to achieve low power audio freq setpoint.

To enable VPU 352MHz feature, select it in menuconfig,
it is disabled by default.

Signed-off-by: Anson Huang <b20788@freescale.com>

Conflicts:
	arch/arm/mach-imx/clk-imx6q.c
	arch/arm/mach-imx/mach-imx6q.c
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/cpufreq/imx6q-cpufreq.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/cpufreq/imx6q-cpufreq.c b/drivers/cpufreq/imx6q-cpufreq.c
index 0ed0c72..85ebcca 100644
--- a/drivers/cpufreq/imx6q-cpufreq.c
+++ b/drivers/cpufreq/imx6q-cpufreq.c
@@ -272,6 +272,13 @@ static void imx6q_opp_check_speed_grading(struct device *dev)
 				dev_warn(dev, "failed to disable 1.2GHz OPP\n");
 	}
 	iounmap(base);
+
+	if (IS_ENABLED(CONFIG_MX6_VPU_352M)) {
+		if (dev_pm_opp_disable(cpu_dev, 396000000))
+			pr_warn("failed to disable 396MHz OPP\n");
+		pr_info("remove 396MHz OPP for VPU running at 352MHz!\n");
+	}
+
 put_node:
 	of_node_put(np);
 }
-- 
1.7.9.5

